src/zap/zapinfo.cpp

```diff
@@ -2086,6 +2104,30 @@ void ZapInfo::GetProfilingHandle(BOOL                      *pbHookFunction,
     *pbIndirectedHandles = TRUE;
 }
 
+DWORD FilterMethodAttribsForIsSupported(DWORD attribs, CORINFO_METHOD_HANDLE ftn, ICorDynamicInfo* pJitInfo)
+{
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+    if (attribs & CORINFO_FLG_JIT_INTRINSIC)
+    {
+        // Do not report the get_IsSupported method as an intrinsic. This will turn the call into a regular
```

Comment:
This is likely going to improve startup but hurt throughput for some of the core functions. That will likely be addressed by rejitting fairly quickly, but its still something to keep in mind.

Long term, if this type of dynamic dispatch is supported, it should likely be changed so that the startup logic caches the CPUID checks and these resolve to be inlined checks of those cached checks (which is how the C Runtime library handles this, for example).

Reply:
Do you have a specific scenario in mind? I can add a benchmark for it (see the benchmark above where I picked one of the affected places). The extra call to a method that returns a constant 1 or 0 didn't hurt perf much.

Reply:
It's likely dependent on how the code was setup and where the ISA check exists. If you have a helper method that was providing a software fallback path (such as `BitOperations.LeadingZeroCount`) and you are calling it in a loop, you are going to have an additional call and branch per loop iteration now which can significantly hurt throughput.

If you just have a centralized check that handles a large block of logic, then the cost of the call is going to be amortized (as was the case for the benchmark above).

Reply:
True, `BitOperations.LeadingZeroCount` in a loop is probably the most pathological case.

I used this to benchmark it:

```csharp
class Program
{
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    static void Main()
    {
        int sum = 0;

        for (int x = 0; x < 10; x++)
        {
            Stopwatch sw = Stopwatch.StartNew();
            for (int i = 0; i < 1000000000; i++)
            {
                sum += BitOperations.LeadingZeroCount((uint)i);
            }
            Console.WriteLine(sw.ElapsedMilliseconds);
        }

        sum.ToString();
    }
}
```

I'm not able to measure a slowdown.

  | Default | TieredCompilation=0
-- | -- | --
After | 511 | 381
Before | 510 | 381

Maybe my CPU is just really good at optimizing that extra call and branch away.

I'm a bit surprised disabling tiering makes this so much faster, given I decorated the method with `AggressiveOptimization`. Cc @kouvel on that.

Reply:
I wasn't able to repro this initially on a recent SDK, but eventually I was able to repro by removing the `AggressiveOptimization` flag and running with `COMPlus_ReadyToRun=0`, even though that doesn't make sense. On my machine at the moment anyway, the numbers are similar to yours but opposite, where it's faster with tiered compilation enabled. The code generated is identical, there doesn't seem to be any alignment issues in the small hot loop, and the method is slightly differently aligned. Looks like in the slow version, there is a decoding stall somewhere, haven't looked into the details.

`AggressiveOptimization` would also prevent prejitting the method such that it gets jitted at runtime with full opts as quickly as possible.